core/trie/snap_support.go

```diff
@@ -0,0 +1,168 @@
+package trie
+
+import (
+	"github.com/NethermindEth/juno/core/felt"
+)
+
+func (t *Trie) IterateAndGenerateProof(startValue *felt.Felt, consumer func(key, value *felt.Felt) (bool, error),
```

Comment:
I'm curious why this approach was taken. Would it not have been simpler to check if there are any leaves to the left (/right) of the startQuery (/endQuery), and return a left (/right) proof? I find the iteration and consumer logic a bit confusing tbh

Reply:
The purpose of this function is to just get the proofs required for the given range and trie right?

Reply:
If we change the function signature to 
`func (t *Trie) IterateAndGenerateProof(startRangeKey, endRangeKey *felt.Felt) ([]ProofNode, bool, error)`
then we could remove the iteration and consumer logic, and simplify this a lot (but just checking if there are any nodes to the left of the startRangeKey, and the right of endRangeKey. What are your thoughts?

Reply:
The purpose of the function is BOTH iterate and generate proof. Think higher level. The reason its a consumer function is that the condition to stop getting leaves is complicated. It need to stop on:
- Context timeout
- Buffer full
- Max node reached
- At least one node found
- Limit node reached
Also, these are relatively large array. Where possible you don't wanna have one []*felt.Felt buffer. EG: in case of class, you actually just want the []core.Class, not the KV, but you do need the proof and you do need the class in the exact order.